# Funciones
Las funciones en Python, y en cualquier lenguaje de programación, son estructuras esenciales de código. Una función es un grupo de instrucciones que constituyen una unidad lógica del programa y **resuelven un problema** muy concreto.

python trae definidas muchas funciones por defecto como:
- print()
- input()
- range()
- len()


y muchas mas (ver [**aqui**](https://www.w3schools.com/python/python_ref_functions.asp)), pero tambien podemos crear nuestras propias funciones.

In [1]:
# ejemplo: funcion que retorna la suma de dos numeros

def suma(a, b):
    return (a+b)

print(suma(3,4))

7


*la funcion suma recibio 2 paramentros (**a, b**) y retorna la suma entre ambos parametros*

## *Return*
La sentencia return es **opcional**, puede devolver, o no, un valor y es posible que aparezca **más de una vez** dentro de una misma función, es decir, de ser necesario **return** devuelve un valor que puede ser almacenado y utilizado en otra parte del programa al terminar la ejecucion de la funcion. si no se usa la sentencia **return** la funcion regresa un valor **None** (*variable sin ningun valor*)

se puede retornar cualquier tipo de **variable** (int, float, str, bool), **colecciones**(listas, tuplas, diccionarios) y tambien varios valores de la siguente manera: **return** valor1, valor2,...

In [2]:
# ejemplo

def operacion_simple(a, b):
    print('elige una operacion:')
    print('1) suma')
    print('2) resta')
    print('3) multiplicacion')
    operacion = int(input('operacion: '))
    if operacion == 1:
        return (a+b)
    elif operacion == 2:
        return (a-b)
    elif operacion == 3:
        return (a*b)
    else:
        print(f'la operacion {operacion} no es valida')

resultado = operacion_simple(4,8)
print('resultado:',resultado)

elige una operacion:
1) suma
2) resta
3) multiplicacion


operacion:  3


resultado: 32


las variables declaradas dentro de la funcion son variables **locales**, es decir, no se pueden utilizar de forma indepeniente en otras partes del codigo.

## Parametros y Argumentos
Es la **informacion** o los **datos** (variables) que recibe la funcion para ejecutarse, los parametros o argumentos son practicamente lo mismo pero los diferenciamos de la siguiente manera:

### Argumentos
se declaran **al momento de crear** la funcion. un funcion puede tener tantos argumentos como se crean necesarios o tambien puede no tener ninguno. y pueden ser utilizados dentro de la funcion.

los argumentos pueden ser de **cualquier tipo** (int, float, str, colecciones, etc)

### Parametros
son los datos que se pasan a la funcion para satisfacer los parametros, **al momento de llamarla**. 

**nota**: si una funcion tiene declarados 3 argumentos estandard, tambien deben pasarsele la misma cantidad de parametros al llamarla, de lo contrario dara error.

In [3]:
# ejemplo

def datos(nombre, edad, adulto):
    #argumentos: nombre, edad, adulto
    print(f'mi nombre es {nombre}, tengo {edad} años, por lo que ya soy adulto... {adulto}')
    
    #parametros
datos('carlos', 24, True)
datos('diana', 17, False)

#si falta alguno de estos parametros daria ERROR

mi nombre es carlos, tengo 24 años, por lo que ya soy adulto... True
mi nombre es diana, tengo 17 años, por lo que ya soy adulto... False


#### Argumentos por defecto
podemos establecer valores por defecto a los argumentos de una funcion, de manera que si no se les asigna un valor (**parametro**) a la hora de llamarla ya tengan un valor definido y por lo tanto no dara error.

In [4]:
# ejemplo
def datos(nombre='S/N', edad = 0):
    print(f'nombre = {nombre}')
    print(f'edad = {edad}')
    
datos()
datos('carlos', 23)

nombre = S/N
edad = 0
nombre = carlos
edad = 23


#### Argumentos variables
si no se sabe con exactitud la cantidad de argumentos que recibira la funcion podemos definir un argumento que recibira una cantidad indefinida de argumentos y se almacenaran en la funcion en forma de **Tupla** (lista inmutable) o **diccionario**

los argumentos pueden ser simples o de clave y valor (**por defecto**) y por convencion se declaran de la seguiente manera:

- argumentos simples: **def** funcion(\*args): ---> **tupla**
- argumentos clave valor: **def** funcion(\*\*kwargs): ---> **diccionario**
- combinacion de ambos: **def** funcion(\*args, \*\*kwargs)

**nota**: los argumentos clave valor (**kwargs** = key word arguments) siempre se declaran de ultimos cuando se crea la funcion en caso de que haya argumentos simples.

In [5]:
# ejemplo 

def nombres(*args):
    # estos elementos no se pueden modificar
    print(args)
    for nombre in args:
        print(nombre)
  
nombres('carlos', 'sebastian', 'angel', 'moises')

('carlos', 'sebastian', 'angel', 'moises')
carlos
sebastian
angel
moises


In [6]:
# ejemplo 2

def terminos(**kwargs):
    print(kwargs)
    for clave, valor in kwargs.items():
        print(f'llave = {clave}')
        print(f'valor = {valor}')
        
    # la clave no lleva comillas 
terminos(nombre='carlos', edad=23, sexo='masculino')

{'nombre': 'carlos', 'edad': 23, 'sexo': 'masculino'}
llave = nombre
valor = carlos
llave = edad
valor = 23
llave = sexo
valor = masculino


In [7]:
#ejemplo 3

# los kwargs se declaran al final
def datos(nombre, *args, **kwargs):
    #el argumento nombre es obligatorio
    print(f'mi nombre es {nombre}', end=" ")
    
    #muestra los nombres ingresados como args
    print(f'tengo {len(args)} hermano(s) ', end="")
    if args:
        for nombre in args:
            print(nombre, end=", ")
            
    #muestra las edades si se ingresan como valores kwargs      
    if kwargs:
        print('de edad: ', end="")
        for edad in kwargs.values():
            print(edad, end=" ")
    print('\n')
            
datos('carlos', 'angel', edad=21)
datos('miguel')
datos('angel', 'aron', 'santiago')

mi nombre es carlos tengo 1 hermano(s) angel, de edad: 21 

mi nombre es miguel tengo 0 hermano(s) 

mi nombre es angel tengo 2 hermano(s) aron, santiago, 



#### Funciones como argumentos
en python una funcion puede recibir otra funcion como argumento.

In [4]:
# ejemplo

# funcion 1
def sumar(a, b):
    return a+b

# funcion 2
def restar(a, b):
    return a-b

# funcion 3
def operacion(funcion, x, y):
    return funcion(x, y)

# llamada a la funcion de operacion y recibe como parametro la funcion sumar 
print(operacion(sumar, 3, 7))
# llamada a la funcion de operacion y recibe como parametro la funcion restar
print(operacion(restar, 17, 3))

10
14


## Tipos de funciones
### Funciones Recursivas
es una funcion que **se llama asi misma** hasta completar una tarea en especifico.

un ejemplo de esto seria calcular el factorial de un numero:

![Imagen](imagenes/recursividad.png)

el factorial de un numero es la multiplicacion **consecutiva** desde 1 hasta el numero del factorial, es decir, el factorial de 5 (**5!**) es igual a multiplicar **1\*2\*3\*4\*5** --> **5! = 1\*2\*3\*4\*5**

sabemos que el factorial de 1 es igual a 1 --> **1! = 1**

tambien que el factorial de un numero es igual multiplicar ese numero por el factorial del numero anterior, es decir: 

**n! = n * (n-1)!** --> **5! = 5 * 4!** asi tambien **4! = 4 * 3!** y asi sucecivamente.

In [3]:
# ejemplo

def factorial(numero):
    if numero == 1:
        # el factorial de 1 es igual a 1 este es el punto de ruptura de la funcion
        return 1
    else:
        # la funcion se llama a si misma hasta que obtenga todos los valores
        return numero*factorial(numero-1)

numero = 5
print(f'el factorial de {numero} es: {factorial(numero)}')

#sin el punto de ruptura la funcion se ejecutaria infinitamente

el factorial de 5 es: 120


la funcion factorial se ejecuta esperando un valor, en este caso **5 \* 4!** pero aun no se ha calculado cuanto vale 4!, entonces deja ese valor en espera y se ejecuta nuevamente, ahora esperando el resultado de **4 \* 3!**, y asi hara en cada iteracion, se llamara asi mismo esperando un resultado hasta que llegue al factorial de 1 (**1!**), aqui el resultado es 1 (**punto de ruptura**) por lo que empezara a **darles valores a todas las llamadas anteriores** hasta la llamada original que realizamos de la funcion y retornar el valor solicitado asi como se ve en la imagen de arriba.

5! = 5\*4! --> 4! = 4\*3! --> 3! = 3\*2! --> 2! = 2\*1! --> 1! = 1 (**punto de ruptura**) 

1! = 1 --> 2! = 2\*1 = 2 --> 3! = 3\*2 = 6 --> 4! = 4\*6 = 24 --> 5! = 5\*24 = 120 (**salida de la funcion**)

### Lambda
son funciones **sencillas** y **anonimas**, es decir que no tienen un nombre y se declaran en **una sola linea de codigo**. Las funciones **lambda** se pueden a una *variable* y tienen la siguiente sintaxis `variable = lambda argumentos: operaciones` o ser llamadas al vuelo `(lambda argumentos: operaciones) (parametros)`

los **argumentos** no se tienen que declarar dentro de parentesis como en las funciones normales y la palabra **return** viene implicita, es decir, no hay que colocarla ya que la funcion lambda retorna automaticamente el resultado de las operaciones que estan a la derecha de los 2 puntos (**:**).

por ultimo las funciones lambda pueden recibir argumentos al igual que las funciones normales:
- sin argumentos
- argumentos simples
- argumentos clave, valor
- \*args
- \*\*kwargs

**nota**: las funciones lambda se escriben en una sola linea por lo que se recomienda para operaciones simples, se pueden usar como funciones anidadas, pasarse directamente como argumentos de otra funcion, etc

In [3]:
#ejemplo 

#funcion tradicional
def sumar(a, b):
    return a + b

# funcion lambda
suma_lambda = lambda a, b: a + b

# llamada a la funcion tradicional
print(sumar(3,7))

# llamada a la funcion lambda
# los parametros se le pasan al igual que a las funciones tradicionales
print(suma_lambda(9,15))

# funcion lambda al vuelo para restar
print((lambda x, y: x-y) (4, 7))

10
24
-3


### Funciones anidadas
son funciones **definidas y llamadas** de otras funciones.

In [8]:
# ejemplo

# funcion principal
def calculadora(a, b, operacion="sumar"):
    
    # funcion anidada 1
    def sumar(a, b):
        return a+b
    
    # funcion anidada 2
    def restar(a,b):
        return a-b
    
    # llama a la funcion anidada sumar
    if operacion == "sumar":
        print('el resultado de la suma es:',sumar(a, b))
    
    # llama a la funcion anidada restar
    elif operacion == "restar":
        print('el resultado de la resta es:',restar(a,b))
        
calculadora(3,5)
calculadora(9,7,'restar')
        

el resultado de la suma es: 8
el resultado de la resta es: 2


### Alcance de las Variables (Scope)
en este ambito hay dos tipos de variables:
- Global
- Local

una variable declarada **fuera de una funcion** se dice que es **Global**, es decir que puede ser accedida o leida desde cualquier parte del programa.

una variable declarada **dentro de una funcion** se dice que es **Local**, es decir que solo puede ser utilizada dentro de la funcion donde se creo o funciones anidadas de esa misma funcion y una vez se termina su ejecucion se destruyen por lo que **ya no pueden ser accedidas fuera de ella**.

una variable global puede ser leida dentro de una funcion pero no puede ser modificada a menos que se anteponga la palabra reservada `global`, esto hace referencia a la variable global y no a una variable local con el mismo nombre dentro de la funcion. tambien puede ser referenciada con la palabra resrvada `nonlocal`

In [18]:
# ejemplo

var_global = 'acceso total'

# funcion principal
def imprimir():
    print(var_global)
    
    var_local = 'acceso solo desde esta funcion o funciones anidadas'
    
    def funcion_anidada():
        # modificando el valor de la variable local de la funcion principal
        nonlocal var_local
        var_local = 'cambio del valor de la variable local de la funcion principal'
        print(var_local)
        
        # modifica el valor de la variable global
        global var_global
        var_global = 'nuevo acceso total'
        print(var_global)
        
        # variable local de la funcion anidada solo se puede usar desde aqui
        var_anidada = 'esta solo existe en la funcion anidada'
        print(var_anidada)
    
    # llamada a la funcion anidada dentro de la funcion principal
    funcion_anidada()

# llamada a la funcion principal
imprimir()
      

acceso total
cambio del valor de la variable local de la funcion principal
nuevo acceso total
esta solo existe en la funcion anidada


### Closure
es una **funcion anidada** que usara las variables locales de la funcion principal o externa y sera **retornada** por la funcion principal. a diferencia de las funciones anidadas la funcion closure **no se llama dentro de la funcion principal**.

In [3]:
# ejemplo

def saludo(nombre):
    # variables locales mensaje y nombre
    mensaje = 'hola ' + nombre
    
    # funcion closure 
    def mostrar_saludo():
        print(mensaje)
        
    # retornando la funcion
    return mostrar_saludo

# llamando a la funcion principal definiendo el parametro de nombre
variable = saludo('Angel')

# llamando a la fucion closure
variable()

hola Angel


In [7]:
# ejemplo 2

# funcion principal
def operacion(a, b):
    
    # funcion lambda anidada : closure
    return lambda: a + b

# llamada a la funcion principal
resultado = operacion(4, 9)
# mostrando la funcion closure
print(f'el resultado es: {resultado()}')

# llamada de la funcion al vuelo
print(f'nueva operacion:{operacion(12,37)()}')


el resultado es: 13
nueva operacion:49


### Decoradores
son funciones que **añaden funcionalidad a otras funciones** sin modificarlas directamente y tienen la siguiente estructura:

 ```
 funcion_a(funcion_b):

     funcion_c():

        nuevas operaciones

        funcion_b()

        nuevas operaciones
     
     return funcion_c
 ```

 - funcion_a = *funcion decoradora*
 - funcion_b = *funcion a decorar*
 - funcion_c = *funcion decorada*

la **funcion_a** (*funcion decoradora*) recibe como parametro una **funcion_b** (*funcion a decorar*) y **retorna** una nueva **funcion_c** (*funcion decorada*) con todas las **funcionalidades u operaciones añadidas**

los decoradores se llaman con la sigiente sintaxis, se llama la funcion decoradora con un (**@**) encima de la funcion a decorar:

```
@funcion_decoradora

def funcion_a_decorar():
```

**nota**: si la funcion a decorar recibe argumentos, deben declararse tambien en la funcion decorada (**funcion_c**), es recomendable que sea como **\*args o \*\*kwargs**

In [11]:
# ejemplo

# declarando funcion decoradora
def funcion_a(funcion_b):
    
    def funcion_c():
        # funcion añadida
        print('ejecutando desde el decorador')
        # llamada de la funcion a decorar
        funcion_b()
        
    # retornando funcion ya decorada
    return funcion_c

# decorando una funcion
@funcion_a
def funcion_a_decorar():
    print('mensaje desde la funcion a decorar')

# llamada de la funcion decorada
funcion_a_decorar()

ejecutando desde el decorador
mensaje desde la funcion a decorar


In [15]:
# ejemplo 2

# funcion decoradora
def mensaje(funcion):
    
    # decorando funcion y agregando argumentos
    def mostrar(*args, **kwargs):
        print('el resultado de la operacion es:')
        resultado = funcion(*args, **kwargs)
        return resultado
    
    # funcion decorada
    return mostrar

# decorando funcion
@mensaje
def sumar(a, b):
    return a+b

# decorando funcion
@mensaje
def potencia(x, y):
    return x**y

# llamando a las funciones decoradas
print(sumar(5, 7))
print(potencia(3, 4))

# las funciones decoradas reciben dos parametros pero gracias al decorador pueden recibir muchos mas 

el resultado de la operacion es:
12
el resultado de la operacion es:
81


### Generadores
son objetos **iterables** (*como listas*) que retornan una secuencia de valores, pero tambien se puede **suspender su ejecucion** por medio de la palabra reservada `yield`.

se pueden acceder a los valores de una funcion generadora por medio de bucles **while** o **for** y tambien por medio de la funcion `next()`.

las funciones generadoras al contrario que las colecciones convencionales **no regresan todos los valores de una vez** en su lugar **van retornando los valores requeridos a medida que se van solicitando**, es decir, **yield** retorna un valor y detiene en ese punto la ejecucion de la funcion **hasta que se pide un valor nuevo**. por lo que en cuestion de uso de memoria son mas eficientes.

cuando ya se hayan retornado todos los valores existentes en el generador no tendra mas valores, por lo que si se vuelve a llamar con `next()` dara error, a menos que se genere el iterador desde el principio.

**nota**: `next()` se utiliza cuando se quieren llamar a los valores generados **uno por uno** o cuando se llaman por el bucle **while**, si se utiliza un bucle **for** no es necesario utilizar `next()`

In [33]:
# ejemplo

def funcion_generadora():
    yield 1 # retorna 1 y suspende la funcion
    
    print('reanudando ejecucion')
    yield 2 # ejecuta el codigo entre ambos yield, retorna 2 y suspende la funcion
    print('reanudando ejecucion nuevamente')
    yield 3

generador = funcion_generadora()

# llamada a la funcion generadora
print(next(generador))

# segunda llamada a la funcion
print(next(generador))

# tercera y ultima llamada a la funcion
print(next(generador))

# si se intenta llamar nuevamente a la funcion generadora dara error ya que no existen mas valores para retornar
# a menos que sea reiniciada

# reiniciando la funcion generadora
generador = funcion_generadora()

# llamando por ciclo for
for valor in generador:
    print(f'desde el for: {valor}')

1
reanudando ejecucion
2
reanudando ejecucion nuevamente
3
desde el for: 1
reanudando ejecucion
desde el for: 2
reanudando ejecucion nuevamente
desde el for: 3


In [34]:
# ejemplo 2

def generador_pares(x):
    for i in range(x):
        if i % 2 == 0:
            yield i
            
generador = generador_pares(10)

for valor in generador:
    print(valor)

0
2
4
6
8


#### Expresiones generadoras
son generadores anonimos parecidos a los lambdas, de esta manera funciona la comprension de listas.

In [37]:
# ejemplo

cuadrados = (valor*valor for valor in range(5))

for i in cuadrados:
    print(i, end=', ')

0, 1, 4, 9, 16, 